contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,308 @@
+     * @param _vault the vault to check.
+     */
+    function _checkIsValidSpread(MarginAccount.Vault memory _vault) internal pure {
+        require(_vault.shortOtokens.length <= 1, "MarginCalculator: Too many short otokens in the vault.");
+        require(_vault.longOtokens.length <= 1, "MarginCalculator: Too many long otokens in the vault.");
+        require(_vault.collateralAssets.length <= 1, "MarginCalculator: Too many collateral assets in the vault.");
+
+        require(
+            _vault.shortOtokens.length == _vault.shortAmounts.length,
+            "MarginCalculator: Short asset and amount mismatch."
+        );
+        require(
+            _vault.longOtokens.length == _vault.longAmounts.length,
+            "MarginCalculator: Long asset and amount mismatch."
+        );
+        require(
+            _vault.collateralAssets.length == _vault.collateralAmounts.length,
+            "MarginCalculator: Collateral asset and amount mismatch."
+        );
+    }
+
+    /**
+     * @dev if there is a short option in the vault, ensure that the long option series being used is a valid margin.
+     * @param _vault the vault to check.
+     */
+    function _checkIsMarginableLong(MarginAccount.Vault memory _vault) internal view {
+        if (_vault.longOtokens.length == 0 || _vault.shortOtokens.length == 0) return;
+
+        OtokenInterface long = OtokenInterface(_vault.longOtokens[0]);
+        OtokenInterface short = OtokenInterface(_vault.shortOtokens[0]);
+        require(_getBatchId(long) == _getBatchId(short), "MarginCalculator: Long and short batch mismatch");
+    }
+
+    /**
+     * @dev internal function to get underlying price of an otoken.
+     * @param _otoken otoken address
+     * @return price the underlying asset price with 18 decimals
+     * @return isFinalized the price is finalized by the oracle and can't be changed
+     */
+    function _getUnderlyingPrice(address _otoken) internal view returns (uint256 price, bool isFinalized) {
+        OtokenInterface otoken = OtokenInterface(_otoken);
+        bytes32 batchId = _getBatchId(otoken);
+        OracleInterface oracle = OracleInterface(AddressBookInterface(addressBook).getOracle());
+        return oracle.getBatchPrice(batchId, otoken.expiryTimestamp());
+    }
+
+    /**
+     * @dev internal function to hash paramters and get batch id. Same batch id means same product with same expiry.
+     * @param _otoken otoken address
+     * @return id the batchDd of an otoken
+     */
+    function _getBatchId(OtokenInterface _otoken) internal view returns (bytes32 id) {
+        id = keccak256(
+            abi.encodePacked(
+                _otoken.underlyingAsset(),
+                _otoken.strikeAsset(),
+                _otoken.collateralAsset(),
+                _otoken.expiryTimestamp()
+            )
+        );
+    }
 
+    function _uint256ToFixedPointInt(uint256 _num) internal pure returns (FixedPointInt256.FixedPointInt memory) {
```

Comment:
This function does not do any scaling right ?

Reply:
I'm assuming that we are scaling to fixed point with 1e18 in the calculator, but all calculation here does not use that.

Reply:
I think this one should be fixed to:
```
function _uint256ToFixedPointInt(uint256 _num) internal pure returns (FixedPointInt256.FixedPointInt memory) {
        return FixedPointInt256.fromUnscaledInt(SignedConverter.uintToInt(_num));
}
```

Reply:
Also we mostly need a function `toUnscaledInt` that scale down before using `intToUint` in any return statement.

Reply:
Resolve this now, will open another issue for this